File: idm/acl/rest/permissions.go
/*
 * Copyright (c) 2018. Abstrium SAS <team (at) pydio.com>
 * This file is part of Pydio Cells.
 *
 * Pydio Cells is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio Cells is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio Cells.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */

package rest

import (
	"context"

	"go.uber.org/zap"
	"google.golang.org/protobuf/types/known/anypb"

	"github.com/pydio/cells/v5/common"
	"github.com/pydio/cells/v5/common/auth"
	"github.com/pydio/cells/v5/common/auth/claim"
	"github.com/pydio/cells/v5/common/client/commons"
	"github.com/pydio/cells/v5/common/client/commons/idmc"
	"github.com/pydio/cells/v5/common/client/commons/treec"
	"github.com/pydio/cells/v5/common/errors"
	"github.com/pydio/cells/v5/common/nodes/abstract"
	"github.com/pydio/cells/v5/common/permissions"
	"github.com/pydio/cells/v5/common/proto/idm"
	"github.com/pydio/cells/v5/common/proto/service"
	"github.com/pydio/cells/v5/common/proto/tree"
	"github.com/pydio/cells/v5/common/telemetry/log"
)

// WriteAllowed returns an error if the Write permission is not present in an acl
func (a *Handler) WriteAllowed(ctx context.Context, acl *idm.ACL) error {

	if claims, ok := ctx.Value(claim.ContextKey).(claim.Claims); ok {
		if claims.Profile == common.PydioProfileAdmin { // Always allow for admins
			return nil
		}
	}

	if acl.NodeID == "" && acl.RoleID != "" {
		// This is a global ACL just used for role
		// Check resource policies associated to this role
		log.Logger(ctx).Debug("Checking acl write on role", zap.Any("acl", acl))
		return a.CheckRole(ctx, acl.RoleID)

	} else if acl.NodeID != "" && acl.Action != nil && (acl.Action.Name == permissions.AclRead.Name || acl.Action.Name == permissions.AclWrite.Name) {
		// Do not bother about roles here, just verify
		// that this node is belonging to an authorized path: use accessList for that
		return a.CheckNode(ctx, acl.NodeID, acl.Action)

	} else {
		log.Logger(ctx).Warn("Cannot check acl right for this request - probably a workspace wide acl delete query - letting it through", zap.Any("acl", acl))
	}

	return nil
}

// CheckRole loads a role and its Policies to check wether this role can be edited
// in the current context
func (a *Handler) CheckRole(ctx context.Context, roleID string) error {

	cli := idmc.RoleServiceClient(ctx)
	q, _ := anypb.New(&idm.RoleSingleQuery{Uuid: []string{roleID}})
	stream, err := cli.SearchRole(ctx, &idm.SearchRoleRequest{Query: &service.Query{SubQueries: []*anypb.Any{q}}})
	var role *idm.Role
	if e := commons.ForEach(stream, err, func(t *idm.SearchRoleResponse) error {
		role = t.GetRole()
		return nil
	}); e != nil {
		return errors.Tag(e, errors.StatusForbidden)
	}
	if role == nil {
		return errors.WithMessagef(errors.StatusForbidden, "role %s does not exists", roleID)
	}
	if !a.MatchPolicies(ctx, role.Uuid, role.Policies, service.ResourcePolicyAction_WRITE) {
		subjects, _ := auth.SubjectsForResourcePolicyQuery(ctx, nil)
		log.Logger(ctx).Debug("Error while checking role from ACL rest : ", zap.Any("role", role), log.DangerouslyZapSmallSlice("subjects", subjects))
		return errors.WithMessage(errors.RoleACLsNotEditable, "while checking role from ACL")
	}
	log.Logger(ctx).Debug("Checking acl write on role PASSED", zap.String("roleId", roleID))
	return nil

}

// CheckNode uses AccessList object to analyze the current ACLs of the target node
func (a *Handler) CheckNode(ctx context.Context, nodeID string, action *idm.ACLAction) error {

	accessList, err := permissions.AccessListFromContextClaims(ctx)
	if err != nil {
		return errors.Tag(err, errors.StatusForbidden)
	}

	treeClient := treec.NodeProviderClient(ctx)
	ancestorStream, lErr := treeClient.ListNodes(ctx, &tree.ListNodesRequest{
		Node:      &tree.Node{Uuid: nodeID},
		Ancestors: true,
	})
	if lErr != nil {
		return lErr
	}
	parentNodes := []*tree.Node{{Uuid: nodeID}}
	for {
		parent, e := ancestorStream.Recv()
		if e != nil {
			if errors.IsStreamFinished(e) {
				break
			}
			if errors.Is(e, errors.NodeNotFound) { // todo - recheck this case
				return nil
			}
			return e
		}
		if parent == nil {
			continue
		}
		parentNodes = append(parentNodes, parent.Node)
	}

	// Update Access List with resolved virtual nodes
	virtualManager := abstract.GetVirtualProvider()
	for _, vNode := range virtualManager.ListNodes(ctx) {
		if aclNodeMask, has := accessList.GetNodesBitmasks()[vNode.Uuid]; has {
			if resolvedRoot, err := virtualManager.ResolveInContext(ctx, vNode, false); err == nil {
				log.Logger(ctx).Debug("Updating Access List with resolved node Uuid", zap.Any("virtual", vNode), zap.Any("resolved", resolvedRoot))
				accessList.AddNodeBitmask(resolvedRoot.Uuid, aclNodeMask)
			}
		}
	}

	var check bool
	if action.GetName() == permissions.AclRead.GetName() {
		log.Logger(ctx).Debug("Checking read on parent nodes", accessList.Zap(), log.DangerouslyZapSmallSlice("parentNodes", parentNodes))
		check = accessList.CanRead(ctx, parentNodes...)
	} else if action.GetName() == permissions.AclWrite.GetName() {
		log.Logger(ctx).Debug("Checking write on parent nodes", accessList.Zap(), log.DangerouslyZapSmallSlice("parentNodes", parentNodes))
		check = accessList.CanWrite(ctx, parentNodes...)
	}

	if !check {
		return errors.WithDetails(errors.WithMessage(errors.StatusForbidden, "you are not authorized to open rights on this node"), "action", action)
	}

	return nil
}

File: frontend/front-srv/assets/action.share/res/js/dialog/links/Permissions.js
import React from 'react';

/*
 * Copyright 2007-2017 Charles du Jeu - Abstrium SAS <team (at) pyd.io>
 * This file is part of Pydio.
 *
 * Pydio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */
import PropTypes from 'prop-types';

import Pydio from 'pydio'
import ShareContextConsumer from '../ShareContextConsumer'
import {Checkbox} from 'material-ui'
import LinkModel from './LinkModel'
import ShareHelper from '../main/ShareHelper'
import {RestShareLinkAccessType} from 'cells-sdk'

export function PermissionsTitle(compositeModel, linkModel, getMessage) {
    let mainString = getMessage('71')
    let perms = [];
    if(linkModel.hasPermission('Preview')){
        perms.push(getMessage('72'))
    }
    if(linkModel.hasPermission('Download')){
        perms.push(getMessage('73'))
    }
    if(linkModel.hasPermission('Upload')){
        if (compositeModel.getNode().isLeaf()){
            perms.push(getMessage('74b'))
        } else {
            perms.push(getMessage('74'))
        }
    }
    if(perms.length){
        mainString = <span>{mainString} <i style={{opacity:.5}}>({perms.join(', ').toLowerCase()})</i></span>
    }
    return mainString
}

class PublicLinkPermissions extends React.Component {
    static propTypes = {
        linkModel: PropTypes.instanceOf(LinkModel),
        style: PropTypes.object
    };

    changePermission = (event) => {
        const name = event.target.name;
        const checked = event.target.checked;
        const {compositeModel, linkModel} = this.props;
        const link = linkModel.getLink();
        if(checked) {
            link.Permissions.push(RestShareLinkAccessType.constructFromObject(name));
        } else {
            link.Permissions = link.Permissions.filter((perm)=>{
                return (perm !== name);
            })
        }
        if(compositeModel.getNode().isLeaf()){
            const auth = ShareHelper.getAuthorizations();
            const max = auth.max_downloads;
            // Readapt template depending on permissions
            if (linkModel.hasPermission('Preview')) {
                link.ViewTemplateName = "pydio_unique_strip";
                link.MaxDownloads = 0; // Clear Max Downloads if Preview enabled
            } else {
                link.ViewTemplateName = "pydio_unique_dl";
                if(max && !link.MaxDownloads) {
                    link.MaxDownloads = max;
                }
            }
        }
        this.props.linkModel.updateLink(link);
    };

    render() {
        const {linkModel, compositeModel, pydio} = this.props;
        const node = compositeModel.getNode();
        let perms = [], previewWarning;
        const auth = ShareHelper.getAuthorizations();

        if(node.isLeaf()){
            const {preview,writeable} = ShareHelper.nodeHasEditor(pydio, node);
            perms.push({
                NAME:'Download',
                LABEL:this.props.getMessage('73'),
                DISABLED:!preview || !linkModel.hasPermission('Preview') // Download Only, cannot edit this
            });
            if(preview && !auth.max_downloads){
                perms.push({
                    NAME:'Preview',
                    LABEL:this.props.getMessage('72'),
                    DISABLED: !linkModel.hasPermission('Download')
                });
                if(linkModel.hasPermission('Preview')){
                    if(writeable){
                        perms.push({
                            NAME:'Upload',
                            LABEL:this.props.getMessage('74b')
                        });
                    }
                }
            }
        } else {
            perms.push({
                NAME:'Preview',
                LABEL:this.props.getMessage('72'),
                DISABLED:!linkModel.hasPermission('Upload')
            });
            perms.push({
                NAME:'Download',
                LABEL:this.props.getMessage('73')
            });
            perms.push({
                NAME:'Upload',
                LABEL:this.props.getMessage('74')
            });
        }

        /*
        if(this.props.shareModel.isPublicLinkPreviewDisabled() && this.props.shareModel.getPublicLinkPermission(linkId, 'read')){
            previewWarning = <div>{this.props.getMessage('195')}</div>;
        }
        */
        return (
            <div>
                {perms.map(function(p){
                    return (
                        <Checkbox
                            key={p.NAME}
                            disabled={p.DISABLED || this.props.isReadonly() || !linkModel.isEditable()}
                            type="checkbox"
                            name={p.NAME}
                            label={p.LABEL}
                            onCheck={this.changePermission}
                            checked={linkModel.hasPermission(p.NAME)}
                            labelStyle={{whiteSpace:'nowrap'}}
                            style={{margin:'10px 0'}}
                        />
                    );
                }.bind(this))}
                {previewWarning}
            </div>
        );
    }
}

PublicLinkPermissions = ShareContextConsumer(PublicLinkPermissions);
export {PublicLinkPermissions as default}